bench/run_benchmarks.rb

```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env ruby
+
+# run_benchmarks.rb
+
+# frozen_string_literal: true
+
+require "benchmark"
+
+require_relative "../lib/regular_expression"
+
+# Individual benchmarks are in the format:
+#     [ pattern, value_to_match, should_match, num_iters_per_batch ]
+
+BENCHMARKS = {
+  "basics" => [
+    ["ab", "ab", true, 10_000],
+    ["ab", "ac", false, 10_000]
+  ]
+}.freeze
+
+NUM_BATCHES = 5
+
+def time_all_matches(source, value, should_match: true, iters_per_batch: 100)
+  samples = {
+    metadata: {
+      source: source,
+      value: value,
+      should_match: should_match,
+      iters_per_batch: iters_per_batch,
+      num_batches: NUM_BATCHES
+    }
+  }
+
+  # Three RE-gem objects, one basic and two compiled
+  begin
+    re_basic, re_x86, re_ruby = *(1..3).map { RegularExpression::Pattern.new(source) }
+    re_x86.compile(compiler: RegularExpression::Compiler::X86)
+    re_ruby.compile(compiler: RegularExpression::Compiler::Ruby)
+  rescue StandardError
+    warn "Exception when building RE objects for regexp (#{source.inspect} / #{value.inspect})"
+    raise
+  end
+
+  [
+    [Regexp.new(source), :ruby, "Ruby built-in regexp"],
+    [re_basic, :re, "uncompiled RegularExpression object"],
+    [re_x86, :re_x86, "RegularExpression x86 compiler"],
+    [re_ruby, :re_ruby, "RegularExpression Ruby compiler"]
+  ].each do |match_obj, samples_name, matcher_description|
+    did_match = match_obj.match?(value)
```

Comment:
This could probably be `!match_object.match?(value).nil?`

Reply:
I don't think so. A no-match returns false, not nil.

Reply:
Or at least, I think I've seen that. Let me go try it.

Reply:
Huh. Ruby built-in regexps are sometimes returning false for regexp match failures. So yeah, "nil?" doesn't work here :-(